### ISSUE TITLE
[Torch Inductor] Torch Inductor Better Support for GNN workload and Inductor Sparse Compiler

### ISSUE THREAD

**Issue by fishmingyu** [pytorch/pytorch#113232]
### 🚀 The feature, motivation and pitch

Torch.compile has been included in Pytorch 2.0 or late versions and has shown that it is a powerful way to speed up Pytorch code greatly. In it, the Pytorch inductor is a new backend that compiles FX graphs generated by TorchDynamo into optimized C++ or Triton kernels. The core loop level intermediate representation (IR) of the inductor only has about 50 operators, and its implementation in Python makes it easily hackable and extensible.   

The Pytorch inductor has pioneered graph-mode execution, significantly accelerating a spectrum of machine learning applications, including CNNs, Transformers, and Graph Neural Networks (GNNs). Despite these advancements, the end-to-end optimization for PyG-based GNNs remains suboptimal. The root of this inefficiency lies in the current implementation of the fused scatter_add kernel, e.g. `fused_index_new_zeros_scatter_add_0`. Also, regarding profiling results, the fused scatter_add kernel has accounted for over 80% of the running time of typical GNN models.  The code provided [here](https://gist.github.com/fishmingyu/60910e86411c60884bdeec3878dfeda3) shows that codegen of fused gather scatter will lower atomic operations, which significantly affects the performance. This proposal sets forth a strategy to refine this mechanism, focusing initially on CPU backends due to their less complex thread hierarchy than GPUs and the current limitations of the Triton backend's warp-level specification. Check here for the [triton limitation](https://github.com/fishmingyu/inductor_test_gather_scatter/tree/main/gpu) for Sparse Kernel Codegen.

## Objective:
The primary goal is to enhance the PyTorch inductor's capability for optimizing scatter-based GNN end-to-end models, specifically targeting CPU backends. The ideal codegen result will be similar to [CSR part of aten::scatter_reduce kernel](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/cpu/ScatterGatherKernel.cpp#L776-L810). 

### Expected Performance:
These enhancements are expected to bridge the gap between the current and optimal performance for GNNs, thereby unlocking new efficiencies and potential within PyTorch's framework. Techniqually. this project will be targeting at:
* Codegen result will achieve the kernel-wise result proposed by [this code](https://github.com/fishmingyu/inductor_test_gather_scatter), where showed the performance speedup from 4-7x.
* End2end result will meet the expectation of 2-3x, considering the [breakdown profiling](https://gist.github.com/fishmingyu/cc339f22869b8ad9a9ffe74aed1c8e22) of mainstream GNN model.

## Techniques:
We will be working on following new ideas:
* Pattern Matching for Scatter-Based Operations:
Investigation reveals that kernels involving scatter operations devolve to atomic-based IRs, which lead to subpar code generation. A pattern matching strategy will be developed to identify and address these inefficiencies. Note that this pattern match happens in `inductor/scheduler.py`. We will add a sparse scheduler pass after fusion process, filtering the scheduler node that need to be optimized. Below is the possible code for reference.
```python 
self.create_foreach_nodes()
self.topological_sort_schedule()
self.logged_slow_fusion = set()
self.fuse_nodes()
self.sparse_aware_nodes() # sparse specific pass
```
* Preprocessing for Sparse Formats:
Recognizing the unique requirements of GNNs, we will implement a detection mechanism for input data indicative of sparsity. Such data will be preprocessed and converted into Compressed Sparse Row (CSR) format, making it more amenable to optimization for sparse operations.
* IR Transformation:
We propose to transform the conventional schedule nodes into new, sparse-specific schedule nodes within the PyTorch inductor's IR. This will facilitate the generation of more efficient execution plans for sparse data. The code below shows a possible IR using inductor's primitive:
```python
buf1.group.device = cpu
buf1.group.iteration = ((nodes, feats), ())
buf1.sizes = ([nodes, feats [])
buf1.mutations = ['buf0']
class buf1_loop_body:
    var_ranges = {z0: nodes, z1: feats}
    index0 = z0
    index1 = z0 + 1
    index2 = feats*indirect1 + z1
    index3 = feats*z0 + z1
    def body(self, ops):
        start_index = self.get_index('index0')
        start = ops.load('csr_ptr', get_index)
        end_index = self.get_index('index1')
        end = ops.load('csr_ptr', get_index)
        out_idx = self.get_index('index3')
        acc = ops.load('C', out_idx) # acc = 0
        for i in range(start, end):
            set_indirect1 = self.set_indirect1(i)
            col_index = self.get_index('indirect1')
            col = ops.load('csr_ind', col_index)
            B_index = self.get_index('index2')
            B_val = ops.load('B_index')
            acc = ops.add(acc, B_val)
        store = ops.store('buf1', out_idx, acc, None)
        return store

```
* Sparse-Aware Code Generation:
Leveraging the modified IR, we will develop a code generation engine capable of producing highly optimized code for executing sparse operations in CSR format on CPU backends. The ideal final code can be checked [here](https://github.com/fishmingyu/inductor_test_gather_scatter/blob/main/openmp/spmm_csr.cpp).

cc @mingfeima @yanbing-j @xinchen9 @zhang677

**Comment by fishmingyu** [comment]
After dicussion with @jansel @aakhundov, I've added some information and perspectives about existing sparse and jagged tensors, as well as the challenges in compiling them. 

### The GNN pattern
It all starts from the kernel fusion of [GAS model](https://github.com/pytorch/pytorch/issues/71300), which is fundamental in graph processing. To be specific, I list the code here:
```python

# Basic "Gather-Apply-Scatter" patterns commonly used in PyG:
def gather_scatter(x, edge_index, reduce="sum"):
    row, col = edge_index
    x_j = x[row]
    return scatter(x_j, col, dim_size=x.size(0), reduce=reduce)

```
The issue that blocks me is that while the GAS model can be fused by the Torch Inductor, it currently reduces to an atomic_add operation. Although various algorithms exist to circumvent atomic_add, in this context, eliminating atomic operations seems feasible only through data transformation, particularly because the edge_index might be unsorted. For example, the edge_index can be created as 
```python
    num_nodes, num_edges = 10_000, 200_000
    edge_index = torch.randint(num_nodes, (2, num_edges), device=args.device)
```
In this code, the `edge_index` indicates that different nodes will be reduced to a single node. However, since we don't know which specific nodes correspond to the reduction, the only way is to use `atomic_add`.

If we do the data transformation in the beginning, for example, we make sure the `edge_index` has been sorted as a input, and then we can transform the GAS model to gather_segment_reduce: (And I think this would be easily handled by PyG's frontend like adding a tag. A code script similar to this principle [here](https://github.com/pyg-team/pytorch_geometric/blob/dadfeec2e38c8a220e478455cbc6d552a8f40ab5/torch_geometric/nn/conv/rgcn_conv.py#L95))

```python
def gather_segment_reduce(x, edge_index, reduce="sum"):
    row, col = edge_index # sorted with col
    x_j = x[row]
    return segment_reduce(x_j, col, reduce=reduce)
```

In this way, since we have sorted the second dimension of edge_index, we can reduce the atomic operations to the largest extent. This is similar to the code brought by torch-scatter's [segment_coo function](https://pytorch-scatter.readthedocs.io/en/latest/functions/segment_coo.html).

However, we found several challenges for the compiler under this scenario:
1. PyTorch's `torch.segment_reduce` function is designed to work with 1D tensors. It performs segment-wise reduction operations based on a corresponding 1D `indices` tensor. If `x_j` is larger than 1-D, e.g. 2-D tensor, this function will fail.
2. `segment_reduce`  can't be fused with gather/index operator, i.e. `x_j = x[row]`.
3. `segment_coo` proposed by PyG isn't compatible with the torch inductor.
4. `segment_reduce` or `segment_coo` is not trival to codegen. Check these links [segment_reduce](https://github.com/pytorch/pytorch/blob/9b0f2f8d94b7485c2c2a4e3d1294a3589b3466b0/aten/src/ATen/native/SegmentReduce.cpp#L32) and [segment_coo](https://github.com/rusty1s/pytorch_scatter/blob/140d3ad677aae615767412873b90982cbf97d35d/csrc/cuda/segment_coo_cuda.cu#L16).



### Jagged Tensor

Jagged tensor is a fundamental data structure, widely used in NLP and recommendation systems (like [DLRM](https://github.com/facebookresearch/dlrm)). 
Here is an example of how to construct jagged tensor by using [torchrec.sparse](https://pytorch.org/torchrec/torchrec.sparse.html):

``` python
values = torch.Tensor([1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0])
offsets = torch.IntTensor([0, 2, 2, 3, 4, 5, 8])
jt = JaggedTensor(values=values, offsets=offsets)
```
To express vividly, we convert it to a jagged view
```
[[1.0, 2.0],
[],
[3.0],
[4.0],
[5.0]
[6.0, 7.0, 8.0]]
```
As we can see, this jagged tensor has a dimension whose slices may be of different lengths. 

### Sparsity within `edge_index` using `torch.Tensor`
Let's create a simple example here
```python
row = torch.Tensor([0, 1, 2, 3, 3, 4, 2, 4, 5, 2])
col = torch.Tensor([1, 2, 3, 1, 5, 2, 4, 3, 3, 1])
edge_index = torch.stack([row, col], dim=0)
```

**How to detect this edge_index is sparse?**
This edge_index can be translated to a sparse matrix, noticing that each column `[0, 1]`, `[1, 2]`, `[2, 3]`, `[3, 1]`, `[3, 5]`, `[1, 2]`, `[2, 4]`, `[4, 3]`, `[5, 3]`, `[2, 1]`, represents an edge in the graph. This format is commonly used in graph neural network frameworks like PyTorch Geometric to represent graph structures. To detect whether it is sparse, we can use the following equation
```latex
sparsity = num_edges / (num_nodes * num_nodes)
```
Usually, when sparity is less than 0.05, we consider the workload here unsuitable for dense representation and will use SpMM rather than GEMM.

Back to the example given by PyG
```python
num_nodes, num_edges = 10_000, 200_000
edge_index = torch.randint(num_nodes, (2, num_edges), device=args.device)
```
The estimated sparsity here is `0.002`, which, though quite small, aligns with the typical range observed in graph problems.
Note here is an estimation since `randint` can't ensure every pair of nodes are mutually different.

### Relationship between jagged tenor and `edge_index`
I appreciate @jansel 's comment about doing things more generically, and I am eager to promote a sparse compiler in a more general way, rather than using domain-specific knowledge. So can we convert `edge_index` to a jagged tensor? The answer is yes.

First of all, let us sort the `edge_index` example I gave above, and I think it can be easily acquired by adding a tag in PyG @rusty1s. Since we need to perform reduction along the column dimension, our `edge_index` needs to be sorted along the column axis. The result should be:
```python
row = torch.Tensor([0, 2, 3, 1, 4, 2, 4, 5, 2, 3])
col = torch.Tensor([1, 1, 1, 2, 2, 3, 3, 3, 4, 5])
edge_index = torch.stack([row, col], dim=0)
```
Then similarly, we can set a jagged tensor, which is indeed a [CSR representation](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.csr_matrix.html)
``` python
row = torch.Tensor([1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0])
offsets = torch.IntTensor([0, 0, 3, 5, 8, 9, 10])
edge_index = JaggedTensor(values=row, offsets=offsets)
```

**How do we understand the sparsity and jagged features?**

From my understanding, sparsity can be broken down into two aspects: Take a sparse matrix as an example. 
* Firstly, the elements in different rows of the matrix are distinct, which essentially reflects its jagged nature. 
* Secondly, the spacing between elements in the same row is irregular and discrete, a characteristic I would describe as 'jumpiness'.

Current hardware, whether CPUs or GPUs, fundamentally cannot address this 'jumpiness' due to their architecture. This is because the issue of 'jumpiness' involves accessing non-continuous memory addresses, a limitation of current architectures. However, resolving the jagged issue already addresses a significant part of the sparsity problem. This is because solving the jagged issue can effectively tackle the atomic operations problem associated with unordered reductions, and these atomic operations are inefficient on both CPUs and GPUs. Therefore, my proposal primarily aims to address this atomic issue.

### What would be the best way to tradeoff hareware effciency, frontend expressiveness and compiler's codegen ability? 

My hope is to incorporate sparsity as a more general aspect within the compiler, rather than as a third-party interface or through a domain-specific language. I hope this feature will be helpful for a range of ML models, including GNN, NLP models and Recommendation systems. Nevertheless, the challenge lies in effectively implementing sparse-aware code generation within the compiler, while optimally balancing performance and development time.

**Comment by bobrenjc93** [comment]
@fishmingyu do you have any updates here? @soulitzer is the NJT work y'all worked on relevant here?

**Comment by fishmingyu** [comment]
Hi @bobrenjc93 , thanks for your interest. Unfortunately, I don't have the bandwidth to engage in this RFC at the moment. However, the details have been discussed in this thread: https://dev-discuss.pytorch.org/t/pytorch-sparse-gnn-compiler-rfc/1644/27.

I’ve also implemented a related Triton kernel, which you can find here: https://github.com/fishmingyu/GeoT/tree/main/geot/triton. The main challenge I encountered was mapping the new IR to Triton codegen within the Inductor backend. If you are familiar with it, maybe we could have a further discussion. Thanks!

**Comment by fishmingyu** [comment]
For the frontend, the only thing to change is to simply add a "sorted" flag in the original scatter_ or index_reduce_ operator. And this flag will be propagated from the frontend to IR and finally to the Triton codegen.
